Rework the assert_return() macro

evelikov opened the pull request:
From the commit introducing it:

```
    The current assert_return() macro has a few short comings:
     - the name implies assert()
     - all users unnecessarily provide EXIT_FAILURE
     - the users cannot provide a message - "Failed assertion:" isn't great

    Tweak it up and in the process bump it to ALL_CAPS, to better indicate
    that it's a macro and not a function.
```

... and the follow-up commit:

```
    Swap the manual checks for the new OK() macro. In the process, we remove
    the cleanup in the error path.

    Many other tests don't bother, plus in the places that do try they don't
    always get it right/fully.
```

Before going forward and porting more tests, would be great to hear if the direction is right.

If* it is, a couple of questions:
 - would people prefer different name - OK one is based on wine, but I'm not married to it ;-)
 - how to handle the conversion - one test at a time, bulk (all assert_return -> OK, all manual -> OK), other?

lucasdemarchi commented:
>     The current assert_return() macro has a few short comings:
>     - the name implies assert()
>    - all users unnecessarily provide EXIT_FAILURE
>    - the users cannot provide a message - "Failed assertion:" isn't great

I think not needing to provide the same EXIT_FAILURE and being able to provide a message is good. The rename I think is not really needed. What's the problem with the term assert?  IMO it's clearer than "ok". igt for example uses a bunch of igt_assert_*.  ok() IMO is odd, but it maybe because I've never seen it used like this. Rust also has both Ok() and assert!(), which I think will add to the confusion.

> how to handle the conversion - one test at a time, bulk (all assert_return -> OK, all manual -> OK), other?

I think a bulk one. If the message is optional, then the first patch is automated. Additional patches may follow later (maybe per compile unit) to add some good messages.

evelikov commented:
@lucasdemarchi humble poke

lucasdemarchi commented:
ts_expect() or ts_assert() would be fine IMO. Other places that use the "expect" terminology:  GoogleTest and kunit... Both of them use semantics like

- KUNIT_EXPECT: Marks the test as failed but continues executing the rest of the test case
- KUNIT_ASSERT: Marks the test as failed and immediately terminates the test case.

So for our uses I think it matches more the "assert" semantics

evelikov committed:
testsuite: introduce TS_ASSERT reimplementing assert_return

The existing macro has a few shortcomings:
 - lowercase, non-prefixed so it can be confused with assert(3), which
   has varying behaviour depending on -DNDEBUG
 - the return value is constant across the project

Introduce TS_ASSERT which is modelled after KUNIT_ASSERT, addressing the
above and tweaking the output format.

Existing instances of assert_return will be updated with a follow-up
commit.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
testsuite: replace assert_return with TS_ASSERT

Mechanical change, done with:

sed -i  "s/assert_return/TS_ASSERT/;s/, EXIT_FAILURE//" testsuite/*.c

In addition, manually fixup the two multiline instances (test-strbuf.c
and test-util.c) and remove the no longer used macro.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
testsuite: don't use assert_return() for empty testsuite

With follow-up commit we'll rework/remove the macro, where the new
version does not allow for providing the return value.

Swap it for a manual ERR, printing a more useful error message.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
testsuite: convert all tests to TS_ASSERT()

Currently we have some inconsistencies across the tests:
 - using TS_ASSERT vs not
 - (non TS_ASSERT tests) leaking on error vs not

In practical terms, we are not too worried about the leaks, since the
test failure comes first. As such, convert all the tests to TS_ASSERT().

This means we loose the useful error messages in a few instances, which
could be re-introduced at a later point alongside a TS_ASSERT_MSG()
macro.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
testsuite: s/return EXIT_SUCCESS/return 0/

With all the EXIT_FAILURE instances done from the tests (only ones in
tools and the test runner remain), lets drop the EXIT_SUCCESS ones.

Functionally identical and arguably slightly neater.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov commented:
Changes:
 - open-code `assert_return(... -EINVAL)` instance
 - `OK` -> `TS_ASSERT`
   - based on `KUNIT_ASSERT`
   - slightly different formatting/reporting - "inspired" by `KUNIT_ASSERT`
   - no user-provided messages - can add `TS_ASSERT_MSG` if needed
 - different split - introduce macro, sed the whole tree
 - convert all tests to `TS_ASSERT`
 - `s/return EXIT_SUCCESS/return 0/` - obligatory OCD bikeshed :sweat_smile:

evelikov commented:
Will check the CI failures later today/tomorrow - the PR should be good for review though.

evelikov commented:
From the failures:
 - arch multilib (clang -m32 failure) - seems like the latest [`lib32-gcc-libs` package has lost `libgcc_s.so`](https://gitlab.archlinux.org/archlinux/packaging/packages/gcc/-/merge_requests/14)
 - style :facepalm: - fixed
 - codecov - temporary unhappy :shrug:

lucasdemarchi commented on testsuite/testsuite.h:
```diff
@@ -119,12 +119,14 @@ int test_run(const struct test *t);
-			    __LINE__, __PRETTY_FUNCTION__);                     \
-			return (r);                                             \
+			ERR("ASSERTION FAILED at %s:%d\n", __FILE__, __LINE__); \
+			ERR("\t" #expr "\n");                                   \
```
From commit message:
   
>  - embeds the expression into the error string, growing the file size

which is still true here, right? Also, for the testsuite the file size due to this would not be of great concern to point out.

evelikov replied:
It meant that the expression string is separate from the boilerplate "foobar failed" error message, allowing the latter to be de-duplicated. Nevertheless, dropped since it's a very meh point.

lucasdemarchi reviewed (approved):
Just a nit on commit message. LGTM.

lucasdemarchi commented:
Applied, thanks
